presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java

```diff
@@ -0,0 +1,220 @@
+                processed[valuePosition] = true;
+                remaining--;
+                summer.writeSum(valueType, valueBlockBuilder, valuePosition, otherValueBlock, otherValuePosition, resultValueBlockBuilder);
+            }
+            else {
+                valueType.appendTo(otherValueBlock, otherValuePosition, resultValueBlockBuilder);
+            }
+        }
+
+        // Now the remaining from orig
+        if (remaining > 0) {
+            Block keyBlock = getKeys();
+            for (int i = 0; i < keyBlock.getPositionCount(); i++) {
+                if (!processed[i]) {
+                    resultKeySet.add(keyBlock, i);
+                    valueType.appendTo(valueBlockBuilder, i, resultValueBlockBuilder);
+                }
+            }
+        }
+
+        return new KeySetAndValues(keyType, valueType, resultKeySet, resultValueBlockBuilder);
+    }
+
+    public void add(Block keyBlock, Block valueBlock, int keyPosition, int valuePosition)
+    {
+        keySet.add(keyBlock, keyPosition);
+        if (valueBlock.isNull(valuePosition)) {
+            if (BIGINT.equals(valueType) || INTEGER.equals(valueType) || REAL.equals(valueType)) {
+                valueType.writeLong(valueBlockBuilder, 0L);
+            }
+            else {
+                valueType.writeDouble(valueBlockBuilder, 0);
+            }
+        }
+        else {
+            valueType.appendTo(valueBlock, valuePosition, valueBlockBuilder);
+        }
+    }
+
+    private static final Summer doubleSummer = new Summer() {
+        @Override
+        public void writeSum(Type type, Block block1, int position1, Block block2, int position2, BlockBuilder blockBuilder)
+        {
+            type.writeDouble(blockBuilder, longBitsToDouble(block1.getLong(position1)) + longBitsToDouble(block2.getLong(position2)));
+        }
+    };
+
+    private static final Summer floatSummer = new Summer() {
+        @Override
+        public void writeSum(Type type, Block block1, int position1, Block block2, int position2, BlockBuilder blockBuilder)
+        {
+            type.writeLong(blockBuilder, Float.floatToIntBits(intBitsToFloat(block1.getInt(position1)) + intBitsToFloat(block2.getInt(position2))));
+        }
+    };
+
+    private static final Summer intSummer = new Summer() {
+        @Override
+        public void writeSum(Type type, Block block1, int position1, Block block2, int position2, BlockBuilder blockBuilder)
+        {
+            type.writeLong(blockBuilder, block1.getInt(position1) + block2.getInt(position2));
+        }
+    };
+
+    private static final Summer longSummer = new Summer() {
+        @Override
+        public void writeSum(Type type, Block block1, int position1, Block block2, int position2, BlockBuilder blockBuilder)
+        {
+            type.writeLong(blockBuilder, block1.getLong(position1) + block2.getLong(position2));
+        }
+    };
+
+    private static interface Summer
```

Comment:
The name sounds really strange to me. What about "aggregator" or "accumulator"?

Reply:
I purposely left it here what people will say lol. Yeah I will change it. It's not aggregator or accumulator. It's an adder. I can call it Adder

Reply:
Done